android/engine/src/main/java/org/smartregister/fhircore/engine/task/FhirTaskUtil.kt

```diff
@@ -108,4 +110,58 @@ constructor(@ApplicationContext val appContext: Context, val defaultRepository:
       ?.outcomeReference
       ?.lastOrNull()
       ?.extractId() == task.logicalId
+
+  suspend fun updateTaskStatuses() {
+    Timber.i("Update tasks statuses")
+
+    val tasks =
+      defaultRepository.fhirEngine.search<Task> {
+        filter(
+          Task.STATUS,
+          { value = of(Task.TaskStatus.REQUESTED.toCoding()) },
+          { value = of(Task.TaskStatus.ACCEPTED.toCoding()) },
+          { value = of(Task.TaskStatus.RECEIVED.toCoding()) },
+        )
+        filter(
+          Task.PERIOD,
+          {
+            prefix = ParamPrefixEnum.LESSTHAN_OR_EQUALS
+            value = of(DateTimeType(Date().plusDays(-1)))
+          }
+        )
+      }
+
+    Timber.i("Found ${tasks.size} tasks to be updated")
+
+    tasks.forEach { task ->
+      // expired tasks are handled by other service i.e. FhirTaskExpireWorker
+      if (task.executionStartIsBeforeOrToday() &&
+          task.status == Task.TaskStatus.REQUESTED &&
+          task.preReqConditionSatisfied()
+      ) {
+        Timber.i("Task ${task.id} marked ready")
+        task.status = Task.TaskStatus.READY
+        defaultRepository.update(task)
+      }
+    }
+  }
+
+  /**
+   * Check in the task is part of another task and if so check if the parent task is
+   * completed,cancelled,failed or entered in error.
+   */
+  private suspend fun Task.preReqConditionSatisfied() =
```

Comment:
I think we can rename this fun and consolidate the precondition

```suggestion
  private suspend fun Task.preconditionSatisfied() =
      this.executionStartIsBeforeOrToday() &&
      this.status == Task.TaskStatus.REQUESTED &&
```